### PR TITLE
H231: Mesh point subsampling TTA — random N-point subset per pass

### PR DESCRIPTION
**Pull request by morganmcg1**
**Current SOTA (PR #1382 H209)**: val_abupt=5.9755, test_abupt=5.8221, test_WSS=6.7214, test_VP=3.4400, test_SP=3.6806
**Source checkpoint**: W&B run `yw2a5dyl` EP13 EMA (`epoch-13`/`best`)
**TTA infra**: `target/eval_tta_h209.py` (read this — it has the mirror_inputs / unmirror_surface_pred / unmirror_volume_pred working baseline; copy and extend)
**Merge gate**: val_abupt < 5.9755 AND test_abupt < 5.8221
**Paper floor**: test_VP ≤ 3.421, test_SP ≤ 3.577, test_WSS ≤ 6.727
**DDP**: 8 GPUs every run
**Budget**: ~45-60min each (eval-only, NO training)

**Context findings just banked (2026-05-29 05:50-06:00Z)**:
- Finding V (askeladd H223): Rotation TTA falsified — H185 has zero rotational invariance at any angle θ ≥ 0.1°
- Finding W (alphonse H224): Coordinate scale TTA falsified — ε=±2% destroys predictions
- Mirror y remains the SOLE validated TTA augmentation for H185

Terminal marker required:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}

## Hypothesis

Findings V + W just falsified geometric perturbation TTA (rotation, scale). All geometric input perturbations destroy because they change the underlying physics problem the model needs to solve. **What hasn't been tested**: input perturbations that DON'T change physics. Mesh point subsampling is one such candidate — different point subsets express the same flow problem at different mesh densities.

**Prediction**: For each TTA pass, randomly subsample 80% of surface_x points and 80% of volume_x points (independent uniform random selection per pass). Average 4 passes. The model has been trained on a fixed mesh density; if it has implicit point-density invariance, the averaged prediction may have reduced variance, gaining 2-5bp like mirror TTA.

If H231 succeeds, stacking with H209 mirror TTA gives orig+mirror+4×subsample = 6-pass TTA — potential SOTA.

## Instructions

Write `target/eval_tta_h231.py` (copy eval_tta_h209.py as starting point). Replace the mirror_inputs/unmirror logic with subsample_inputs:

- For each pass i, randomly select 80% of surface_x indices, 80% of volume_x indices (independent, uniform random)
- Run forward pass on the subset
- The output predictions are at the subset locations — need to evaluate on the FULL eval set for fair comparison
- **Two reasonable strategies for averaging**:
  a) Predict on the full input each pass but ZERO OUT 20% of attention contributions (mask 20% of input points as inactive, predict full output). This keeps output points consistent across passes.
  b) Predict on each subset, then for each full-eval target point, average predictions from passes where it was included (Monte Carlo coverage).

  Use strategy (a) since it gives complete prediction at all output points every pass.

Run modes:
1. orig (no augmentation) — sanity, should match 6.0172/5.8613
2. subsample_only — 4 passes of 80%-masked inputs, 0.25 average
3. mirror + subsample_combined — 5 passes: orig + mirror + 3×subsample, mean

```bash
torchrun --standalone --nproc-per-node=8 target/eval_tta_h231.py \
  --checkpoint outputs/h185_ep13/checkpoint.pt \
  --eval-modes original,subsample_only,mirror_subsample \
  --subsample-rate 0.8 \
  --num-subsample-passes 4 \
  --output-dir outputs/h231_eval \
  --wandb-group h231-askeladd-mesh-subsample-tta \
  --wandb-name "askeladd/h231-mesh-subsample-tta" \
  --batch-size 2
```

If subsample fails catastrophically (val > 7%), bonus arm: try 95% retention (smaller perturbation). If still fails → another "Finding X" landed: mesh sampling not invariant either, mirror is uniquely preserved.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:

Started: 2026-05-29 ~06:30Z. Implementation `eval_tta_h231.py`. PR-spec strategy (a) naive mean + bonus strategy (b) Monte-Carlo coverage included since both share the same forward passes.

**Implementation note on strategy (a)**: The model applies `surface_preds = surface_out(hidden) * surface_mask` and zeros features mid-forward at masked positions. With 4 subsample passes at keep_rate=0.8, the naive /4 mean is biased toward zero by ~20% at each point (`E[pred] = 0.8 * true_pred`). MC-coverage (`sum / count_active`) corrects this bias. I'm running both so we can disentangle the architecture-induced bias from genuine model degradation under subsampling.

**Debug-mode smoke test (1 GPU, 2 cases, 8192 pts/case, 2 sub passes)**:
- `original`: val_abupt=5.44 / test_abupt=6.92
- `subsample_only_naive`: val=30.91 / test=31.13 — catastrophic
- `subsample_only_mc`: val=18.39 / test=19.02 — better than naive (bias removed) but still catastrophic

**Caveat**: H185 was trained on 65536 points/case. Debug mode subsamples 80% of 8192=6553 points, putting the model very far OOD. Full-resolution eval at 65536 → 52428 points may still be in-distribution. Launching the full 8-GPU eval now to confirm.

**Phase 1 launched (8 GPU torchrun)**:
- All 6 modes (original, mirror_only, subsample_only_naive, subsample_only_mc, mirror_subsample_naive, mirror_subsample_mc)
- subsample_rate=0.8, num_passes=4, batch_size=2
- W&B: `askeladd/h231-sub80-4pass` (group h231-askeladd-mesh-subsample-tta)
- Expected runtime: ~150 min (6 fwd passes per batch)

If 0.8 subsample fails as expected, will run Phase 2 bonus arm at keep_rate=0.95.

**Comment by morganmcg1** [comment]
STUDENT askeladd:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["4goypilw","ydoa3bum"],"primary_metric":{"name":"val_abupt_axis_mean_rel_l2_pct","value":5.9832},"test_metric":{"name":"test_abupt_axis_mean_rel_l2_pct","value":5.8289}}

## Results

**H231 FALSIFIED** — Mesh point subsampling TTA does NOT improve over the H209 `mirror_tta` SOTA at any tested keep_rate. Stacking subsample with mirror is strictly worse than mirror alone.

### Merge gate check
- Gate: `val_abupt < 5.9755 AND test_abupt < 5.8221`
- Best H231 variant: `mirror_subsample_mc @ keep=0.8` → val=5.9832 / test=5.8289 — **fails gate on both metrics**

### Phase 1 — `subsample_rate=0.80`, 4 sub passes (W&B `4goypilw`)

| mode                    | val_abupt | test_abupt | val_SP  | test_SP | val_WSS | test_WSS | val_VP  | test_VP |
|-------------------------|-----------|------------|---------|---------|---------|----------|---------|---------|
| original                | 6.0172    | 5.8613     | 3.9673  | 3.7046  | 6.8104  | 6.7642   | 3.5412  | 3.4584  |
| mirror_only             | 6.0195    | 5.8638     | 3.9686  | 3.7068  | 6.8123  | 6.7671   | 3.5468  | 3.4573  |
| subsample_only_naive    | 25.4058   | 25.3826    | 21.6656 | 21.6697 | 26.3895 | 26.4703  | 22.2700 | 22.2950 |
| subsample_only_mc       | 7.0267    | 6.8950     | 5.0358  | 4.8193  | 7.7324  | 7.7005   | 4.7319  | 4.6822  |
| mirror_subsample_naive  | 17.0431   | 16.9686    | 14.3005 | 14.2210 | 17.7714 | 17.7890  | 14.6792 | 14.6659 |
| **mirror_subsample_mc** | **5.9832**| **5.8289** | 3.9467  | 3.6845  | 6.7731  | 6.7293   | 3.5279  | 3.4413  |

### Phase 2 — `subsample_rate=0.95`, 4 sub passes (W&B `ydoa3bum`)

| mode                    | val_abupt | test_abupt | val_SP  | test_SP | val_WSS | test_WSS | val_VP  | test_VP |
|-------------------------|-----------|------------|---------|---------|---------|----------|---------|---------|
| original                | 6.0172    | 5.8613     | 3.9673  | 3.7046  | 6.8104  | 6.7642   | 3.5412  | 3.4584  |
| mirror_only             | 6.0195    | 5.8638     | 3.9686  | 3.7068  | 6.8123  | 6.7671   | 3.5468  | 3.4573  |
| mirror_tta (H209 SOTA)  | **5.9755**| **5.8221** | 3.9423  | 3.6806  | 6.7645  | 6.7214   | 3.5261  | 3.4400  |
| subsample_only_mc       | 6.0148    | 5.8588     | 3.9675  | 3.7049  | 6.8071  | 6.7609   | 3.5425  | 3.4585  |
| mirror_subsample_mc     | 5.9850    | 5.8309     | 3.9476  | 3.6857  | 6.7752  | 6.7313   | 3.5276  | 3.4429  |

### Commands

```bash
# Phase 1 (sub80)
torchrun --standalone --nproc-per-node=8 eval_tta_h231.py \
  --checkpoint outputs/h185_ep13/checkpoint.pt \
  --eval-modes original,mirror_only,subsample_only_naive,subsample_only_mc,mirror_subsample_naive,mirror_subsample_mc \
  --subsample-rate 0.8 --num-subsample-passes 4 \
  --output-dir outputs/h231_eval_sub80 \
  --wandb-name "askeladd/h231-sub80-4pass" \
  --batch-size 2

# Phase 2 (sub95) — includes mirror_tta for direct SOTA comparison
torchrun --standalone --nproc-per-node=8 eval_tta_h231.py \
  --checkpoint outputs/h185_ep13/checkpoint.pt \
  --eval-modes original,mirror_only,mirror_tta,subsample_only_mc,mirror_subsample_mc \
  --subsample-rate 0.95 --num-subsample-passes 4 \
  --output-dir outputs/h231_eval_sub95 \
  --wandb-name "askeladd/h231-sub95-4pass" \
  --batch-size 2
```

- Runtime: ~16 min per phase (387s val + 587s test = ~16.2 min each)
- W&B runs: `4goypilw` (sub80), `ydoa3bum` (sub95)
- Peak memory: not instrumented this run; H185 eval @ 8×GPU bs=2 with 6 fwd passes/batch fits comfortably on 80GB H100 (eval-only, no optimizer state)

### What happened — analysis

**1. Architecture-induced bias (naive vs MC).** The model applies `surface_preds *= surface_mask` mid-forward and zeros features at masked positions. With 4 keep=0.8 passes, the naive `/N` mean is biased toward zero (`E[pred] ≈ keep_rate × true`), producing the catastrophic 25.4% and 17.0% errors. The MC-coverage estimator (`sum / count_active`) recovers an unbiased per-point prediction, confirming the >25% naive numbers are entirely an averaging artifact, not actual model degradation.

**2. Subsample alone is not a useful TTA.** At keep=0.95 (mild perturbation), `subsample_only_mc` gives val=6.0148 / test=5.8588 — a 0.002–0.003bp improvement over `original` (6.0172 / 5.8613). At keep=0.80, even with MC correction, val=7.0267 / test=6.8950 — a 17–18% relative degradation. **The model has very weak point-density invariance**: removing even 5% of mesh points perturbs predictions just enough to lose any TTA-style variance reduction, and 20% removal is a clear OOD shift.

**3. Stacking with mirror is strictly harmful.** At keep=0.95 (the most benign setting), `mirror_subsample_mc` gives val=5.9850 / test=5.8309 vs `mirror_tta` val=5.9755 / test=5.8221 — adding subsample passes to the mirror ensemble **erodes the mirror TTA gain**. The subsample passes inject more noise than mirror_tta gains from variance reduction. At keep=0.80, the best stacked result (val=5.9832 / test=5.8289) still loses to mirror_tta alone.

**4. Why this likely fails for H185.** H185 was trained on a fixed 65536 surface / 65536 volume point density. Slice attention with 128 learned slice anchors implicitly memorizes a point-density profile during training (slice weights are computed from per-point features and softmax-normalized across all points). Subsampling perturbs both the per-slice point counts and the geometric coverage of each anchor, breaking the learned alignment in a way mirror reflection (which preserves all points and densities) does not.

### Finding X candidate

H231 falsified joins H223 (rotation, askeladd) and H224 (coordinate scale, alphonse) as a third negative TTA result on H185:

> **Finding X (provisional):** Mesh point subsampling is not a useful TTA for H185. Both 5% and 20% subsampling actively degrade predictions even after Monte-Carlo coverage correction, and stacking with mirror_tta strictly reduces the mirror gain. Mirror-y remains the **sole validated input-space TTA** for the H185 SOTA. The cluster of falsifications (rotation, scale, subsample) is consistent with H185 being highly tuned to its exact training-point distribution and geometry, with mirror_y being the only true symmetry the model has internalized.

### Suggested follow-ups

- **Output-space TTA (untried)**: rather than perturbing inputs, perturb the *prediction reconstruction*. e.g. predict at a fixed resolution and average predictions queried at slightly shifted target points, or use the model as a function evaluator and compute expected values over small target-point neighborhoods.
- **Slice-anchor stochasticity**: rather than subsampling inputs, perturb the slice-attention assignment (re-roll slice anchors during TTA). This would test invariance to the learned slice partition rather than the input mesh.
- **Checkpoint-ensemble TTA**: average predictions from multiple training checkpoints (e.g. SWA-style mean of EP10/11/12/13) instead of input perturbations. This is the most promising untested direction — uses model variability rather than input variability, which the falsification cluster suggests is the more robust axis.
- **Stop pure-TTA search**: 3 falsifications strongly suggest the input-space TTA budget is exhausted at H185. Future TTA work should explore prediction-space averaging, multi-checkpoint ensembling, or accept that mirror-only is the ceiling for this architecture.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1404 because ADVISOR 2026-05-29 07:55Z — close + Finding AA banked.

**Verdict**: No merge — best result mirror_subsample_mc val=5.9832/test=5.8289 fails both OLD H209 gate (5.9755/5.8221) and NEW H236 gate (5.9613/5.8081). H236 multi-res TTA just merged as new SOTA.

**Finding AA (program-permanent)**: Mesh point subsampling TTA is FALSIFIED for H185 EP13. At keep=0.8 (80% retention), subsample passes actively degrade predictions even after MC-coverage bias correction (val 7.03). Stacking with mirror strictly harms the mirror TTA signal (val mirror+subsample=5.9832 vs mirror alone=5.9755). At keep=0.95 (mild), subsample-only adds only ~2bp improvement (noise floor), and mirror+subsample still fails to beat mirror alone (5.9850 vs 5.9755). Model has no point-density invariance — slice attention learns from full-density mesh implicitly and subsampled inputs are OOD.

**Mechanism confirmed**: surface_preds *= surface_mask zeros outputs at subsampled-out points. Naive /K average biased toward zero by (1-keep_rate). MC-coverage correction removes bias but cannot recover lost structural information.

**NEW SOTA context**: H236 multi-res TTA (nezuko, PR #1408, merged) demonstrated that DIFFERENT volume point counts (49k/65k/82k) do give useful signal — the key is that they're still full-mesh at each resolution, not a subset. Your negative result here perfectly complements that positive result: changing WHICH points you use (subsample) is harmful; changing HOW MANY slices you use per case (resolution/stride) is beneficial.

**Next assignment incoming**: H243 — extended multi-res TTA on H185 using a WIDER resolution range {32768, 49152, 65536, 81920, 98304}×{orig, mirror} = 10-pass TTA. Uses nezuko's eval_multi_res.py directly. ~25min eval-only. If more resolution points push past 5.9613 val, that's a new SOTA..
